cairo/ethereum/cancun/vm/instructions/storage.cairo

```diff
@@ -95,3 +96,127 @@ func sload{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, poseidon_ptr: Poseidon
     let ok = cast(0, ExceptionalHalt*);
     return ok;
 }
+
+// @notice Stores a value at a certain key in the current context's storage.
+func sstore{
+    range_check_ptr, bitwise_ptr: BitwiseBuiltin*, poseidon_ptr: PoseidonBuiltin*, evm: Evm
+}() -> ExceptionalHalt* {
+    alloc_locals;
+    // STACK
+    let stack = evm.value.stack;
+    with stack {
+        let (key, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+        let (new_value, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+    }
+
+    let is_gas_left_not_enough = is_le(evm.value.gas_left.value, GasConstants.GAS_CALL_STIPEND);
+    if (is_gas_left_not_enough != 0) {
+        tempvar err = new ExceptionalHalt(OutOfGasError);
+        return err;
+    }
+
+    // Get storage values
+    let key_bytes32 = U256_to_be_bytes(key);
+    let state = evm.value.env.value.state;
+    let current_target = evm.value.message.value.current_target;
+    with state {
+        let original_value = get_storage_original(current_target, key_bytes32);
+        let current_value = get_storage(current_target, key_bytes32);
+    }
+
+    // Gas calculation
+    // Check accessed storage keys
+    tempvar accessed_tuple = TupleAddressBytes32(
+        new TupleAddressBytes32Struct(current_target, key_bytes32)
+    );
+    let (serialized_keys: felt*) = alloc();
+    assert serialized_keys[0] = accessed_tuple.value.address.value;
+    assert serialized_keys[1] = accessed_tuple.value.bytes32.value.low;
+    assert serialized_keys[2] = accessed_tuple.value.bytes32.value.high;
+    let dict_ptr = cast(evm.value.accessed_storage_keys.value.dict_ptr, DictAccess*);
+    with dict_ptr {
+        let (is_present) = hashdict_read(3, serialized_keys);
+        if (is_present == 0) {
+            hashdict_write(3, serialized_keys, 1);
+            tempvar gas_cost = GasConstants.GAS_COLD_SLOAD;
+            tempvar poseidon_ptr = poseidon_ptr;
+            tempvar dict_ptr = dict_ptr;
+        } else {
+            tempvar gas_cost = 0;
+            tempvar poseidon_ptr = poseidon_ptr;
+            tempvar dict_ptr = dict_ptr;
+        }
+    }
+    let gas_cost = [ap - 3];
+    let poseidon_ptr = cast([ap - 2], PoseidonBuiltin*);
+    let dict_ptr = cast([ap - 1], DictAccess*);
+
+    // Calculate storage gas cost
+    if (original_value.value == current_value.value) {
+        if (current_value.value != new_value.value) {
+            if (original_value.value == 0) {
+                tempvar gas_cost = gas_cost + GasConstants.GAS_STORAGE_SET;
+            } else {
+                tempvar gas_cost = gas_cost + (
+                    GasConstants.GAS_STORAGE_UPDATE - GasConstants.GAS_COLD_SLOAD
+                );
+            }
+        }
+    } else {
+        tempvar gas_cost = GasConstants.GAS_WARM_ACCESS;
+    }
```

Comment:
```suggestion
if (original_value.value == current_value.value and current_value.value != new_value.value) {
        if (original_value.value == 0) {
            tempvar gas_cost = gas_cost + GasConstants.GAS_STORAGE_SET;
        } else {
            tempvar gas_cost = gas_cost + (
                GasConstants.GAS_STORAGE_UPDATE - GasConstants.GAS_COLD_SLOAD
            );
        }
    } else {
        tempvar gas_cost = gas_cost + GasConstants.GAS_WARM_ACCESS;
    }
```

Use of and should work ?

Reply:
i am copying eels here

Reply:
https://github.com/ethereum/execution-specs/blob/master/src/ethereum/cancun/vm/instructions/storage.py#L94-L100
```python
    if original_value == current_value and current_value != new_value:
        if original_value == 0:
            gas_cost += GAS_STORAGE_SET
        else:
            gas_cost += GAS_STORAGE_UPDATE - GAS_COLD_SLOAD
    else:
        gas_cost += GAS_WARM_ACCESS
```

Reply:
ah ok sorry yes, misunderstood

Reply:
so it turns out you can not do (if A and B){ } __ELSE__ {}

Reply:
interesting 
thanks for looking into it